Making batching rule for F.embedding DTensor-aware

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #162117

`vmap(F.embedding)(DTensor, DTensor)` was failing because F.embedding's
batching rule generates a new tensor via at::arange, at::arange
generates a regular tensor, and DTensor rightfully errors on mixed
DTensor-regular Tensor operations.

This PR fixes the problem by activating DTensor implicit replication on
just the at::arange and the subsequent add operation.

In order to accomplish this I move the DTensor implicit replication flag
to C++ (most batching rules are in C++).

Test Plan:
- new test

cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta @ezyang @msaroufim

zou3519 committed:
Batching rules assume DTensor implicit replication

`vmap(F.embedding)(DTensor, DTensor)` was failing because F.embedding's
batching rule generates a new tensor via at::arange, at::arange
generates a regular tensor, and DTensor rightfully errors on mixed
DTensor-regular Tensor operations.

This PR fixes the problem by activating DTensor implicit replication on
every non-randomness batching rule. This is safe to do, because batching
rules must return tensors of the same shape and factory functions will
not return tensors of different values.

In order to accomplish this I move the DTensor implicit replication flag
to C++ (most batching rules are in C++).

Test Plan:
- new test

[ghstack-poisoned]

zou3519 committed:
Update on "Batching rules assume DTensor implicit replication"


`vmap(F.embedding)(DTensor, DTensor)` was failing because F.embedding's
batching rule generates a new tensor via at::arange, at::arange
generates a regular tensor, and DTensor rightfully errors on mixed
DTensor-regular Tensor operations.

This PR fixes the problem by activating DTensor implicit replication on
every non-randomness batching rule. This is safe to do, because batching
rules must return tensors of the same shape and factory functions will
not return tensors of different values.

In order to accomplish this I move the DTensor implicit replication flag
to C++ (most batching rules are in C++).

Test Plan:
- new test

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta

[ghstack-poisoned]

zou3519 commented on torchgen/gen_vmap_plumbing.py:
```diff
@@ -238,10 +242,28 @@ def gen_vmap_plumbing(native_function: NativeFunction) -> str | None:
 template <typename batch_rule_t, batch_rule_t batch_rule>
 {sig.decl(name=schema.name.unambiguous_name() + "_generated_plumbing")} {{
   c10::impl::ExcludeDispatchKeyGuard guard(DispatchKey::FuncTorchBatched);
+  at::DTensorAllowImplicitReplication guard2;
```
Brian had the same question with an example and I see the problem now

bdhirsh commented on torchgen/gen_vmap_plumbing.py:
```diff
@@ -238,10 +242,28 @@ def gen_vmap_plumbing(native_function: NativeFunction) -> str | None:
     wrapped_returns = gen_returns(returns, cur_level_var, results_var)
+    # NOTE: [vmap and DTensor Implicit Replication]
+    # For all batching rules, we assume that new tensors generated by factory
+    # functions are meant to be replicated by DTensor. This saves us
```
> The main thing that I think can go wrong is if you have BatchedTensor(OtherSubclass(DTensor)) as indices_ and somehow the OtherSubclass's torch_dispatch is wrong and causes a mixed DTensor-Tensor operation.

Hmm. I guess can interpose between the call to `aten.add(torch.Tensor, OtherSubclass(DTensor))` and do "anything". So the bad outcome is something like... `OtherSubclass` constructs its own temporary tensor that we know should *not* be replicated as part of intercepting `aten.add`, but we have implicit replicate turned on during the entire `aten.add` call.

I agree that there is some risk, although I can't think of any way to do this that isn't extremely synthetic (on top of that - in practice, we have only ever seen cases of `DTensor(InnerSubclass)`, not the other way around).

zou3519 committed:
Update on "Making batching rule for F.embedding DTensor-aware"


`vmap(F.embedding)(DTensor, DTensor)` was failing because F.embedding's
batching rule generates a new tensor via at::arange, at::arange
generates a regular tensor, and DTensor rightfully errors on mixed
DTensor-regular Tensor operations.

This PR fixes the problem by activating DTensor implicit replication on
just the at::arange and the subsequent add operation.

In order to accomplish this I move the DTensor implicit replication flag
to C++ (most batching rules are in C++).

Test Plan:
- new test

[ghstack-poisoned]

bdhirsh commented on build_variables.bzl:
```diff
@@ -1087,6 +1087,7 @@ aten_cpu_source_non_codegen_list = [
     "aten/src/ATen/DeviceAccelerator.cpp",
     "aten/src/ATen/Context.cpp",
     "aten/src/ATen/DLConvertor.cpp",
+    "aten/src/ATen/DTensorState.cpp",
```
oh god non-globbed ATen file sources

bdhirsh reviewed (approved):
sgtm!
